backports 1.1: Fix namespace package declaration

skupr-anaconda opened the pull request:
**Destination channel:** main

### Links

- [PKG-5442](https://anaconda.atlassian.net/browse/PKG-5442) 
- [Upstream repository](https://github.com/brandon-rhodes/backports/tree/v1.1)

### Explanation of changes:

- Add feedstock's license and README.md
- Replace `__init__.py` with a new `init.py`; remove `build.sh`, `bld.bat` and add `build/script` to `meta.yaml` (taken from the conda-forge's PR https://github.com/conda-forge/backports-feedstock/pull/7)
- Use `set -ex` and `echo on` in `build/scripts` for printing command executions to _stdout_
- Check `{{ PYTHON }} -c "import backports"` in `build/script`
- Add tests from the conda-forge's PR https://github.com/conda-forge/backports-feedstock/pull/7 but extend it by checking all `backports` packages available in _defaults_ 

### Notes:

- Rebuilding because  `setuptools >=70.0.0` can break `backports` packages (see https://github.com/AnacondaRecipes/setuptools-feedstock/pull/62)

[PKG-5442]: https://anaconda.atlassian.net/browse/PKG-5442?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

skupr-anaconda committed:
Fix namespace package declaration

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_6e_al49fyn/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:67: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

skupr-anaconda committed:
Enable echo on win too

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_bdjry0_l21/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:68: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

skupr-anaconda committed:
Enable echo on for tests too

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_1b86gk92mn/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:69: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

ifitchet commented on recipe/meta.yaml:
```diff
@@ -2,16 +2,23 @@
 build:
-  number: 0
+  number: 1
   noarch: python
```
Do we want the `noarch` line?

skupr-anaconda replied:
Yes, because we bump the build number.  For a new version we would remove it

ifitchet commented on recipe/meta.yaml:
```diff
@@ -2,16 +2,23 @@
+  script:
+    - set -ex  # [unix]
+    - echo on  # [win]
+    - cd {{ SP_DIR if SP_DIR is defined else '' }}
```
I think you want your default directory to be `'.'`. `cd` with no arguments on Unix will take you to your `HOME` directory (not sure on Windows) which is not the build directory.

skupr-anaconda replied:
I'll simplify it to `cd {{ SP_DIR }}`

ifitchet replied:
Hmm, that'll still result in `cd` if `SP_DIR` is (ever) unset.  In this case we can probably rely on `SP_DIR` being set as it is a documented variable.

(You'd like to think we'd get an error if we used unset variables.)

ifitchet commented on recipe/meta.yaml:
```diff
@@ -25,8 +32,32 @@ test:
   commands:
     - pip check
+    - echo on                                   # [win]
+    - unset PIP_NO_INDEX                        # [unix]
+    - set "PIP_NO_INDEX="                       # [win]
```
Is it obvious why we don't want `PIP_NO_INDEX` set here or would a comment help?

skupr-anaconda replied:
I'm unsure what was behind the decision to use this env variable. I added a comment why it's here

ifitchet commented on recipe/meta.yaml:
```diff
@@ -25,8 +32,32 @@ test:
+    - unset PIP_NO_INDEX                        # [unix]
+    - set "PIP_NO_INDEX="                       # [win]
+    - python -m pip install --target alt-site backports.tarfile
+    - export PYTHONPATH="${SP_DIR}:alt-site"    # [unix]
+    - set "PYTHONPATH=%SP_DIR%;alt-site"        # [win]
```
What does `alt-site` do?  Maybe another comment unless it it a well-known Python thing that Google isn't telling me about.

skupr-anaconda replied:
`alt-site` is a target name from `python -m pip install --target alt-site backports.tarfile`. 
It will make the PYTHONPATH equal to `export PYTHONPATH=/croot/backports_1722601682062/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.8/site-packages:alt-site`

skupr-anaconda replied:
I guess it's a kind of isolation if I understand well this comment https://github.com/conda-forge/backports-feedstock/pull/7#discussion_r1691594258

skupr-anaconda replied:
Added a comment

ifitchet replied:
Maybe?

"... ahead of alt-site, the value of --target in the above install command"

skupr-anaconda replied:
Fixed

ifitchet commented:
Overall, the PR description isn't telling me _why_ we're doing this, nor does the Jira ticket.

skupr-anaconda committed:
Address comments

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_78dmunnqkc/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:72: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

skupr-anaconda commented:
> * ebuilding because  `setuptools >=70.0.0` can break `backports` packages

@ifitchet I addressed your comments and clarified _why_ we rebuild this package

skupr-anaconda committed:
Add a comment about init.py

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_c7hj_fad20/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:75: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

skupr-anaconda committed:
Improve a comment

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_44s21a_rf9/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:75: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

ifitchet reviewed (approved):
_Now_ I know what going on, LGTM!

(You may still want someone with more clue...)

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_7ccp849m2n/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:75: missing_license_file: The recipe is missing the `about/license_file` or `about/license_url` key.
===== Final Report: =====
0 Errors and 1 Warning were found

</details>
